src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts

```diff
@@ -907,103 +1022,13 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 					eventCounter = 0;
 					this._fireTaskEvent(TaskEvent.create(TaskEventKind.End, task));
 					toDispose.dispose();
+					terminal?.write(`\x1b]633;D;${exitCode}\x07`);
```

Comment:
What if `waitOnExit` was a `boolean | string | (exitCode) => string` to handle the exit code??

Reply:
`waitOnExit` happens before the task has output

Reply:
`waitOnExit` should be written when the process exits, so if we make that a callback instead of a static string we can inject it at that point like this:

```ts
function waitOnExitCallback(message: string): (exitCode: string) => string {
  return (exitCode) => {
    return `\x1b]633;D;${exitCode}\x07${message}`;
  };
}

waitOnExit = waitOnExitCallback(nls.localize('closeTerminal', 'Press any key to close the terminal.'));
```

Then handle the function case here:

https://github.com/microsoft/vscode/blob/18408b909006d18e2e08e8b1c50d20b31854b644/src/vs/workbench/contrib/terminal/browser/terminalInstance.ts#L1629-L1631

Something like this:

```ts
				switch (typeof this._shellLaunchConfig.waitOnExit) {
					case 'string':
						xterm.raw.write(formatMessageForTerminal(this._shellLaunchConfig.waitOnExit, { excludeLeadingNewLine: true }));
						break;
					case 'function':
						xterm.raw.write(formatMessageForTerminal(this._shellLaunchConfig.waitOnExit(exitCode), { excludeLeadingNewLine: true }));
						break;
				}
```

Reply:
clever 😄